pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go

```diff
@@ -468,26 +515,60 @@ func (a *index) AddressInformation(addresses sets.String) ([]model.AddressInfo,
 }
 
 func (a *index) ServicesForWaypoint(key model.WaypointKey) []model.ServiceInfo {
-	var out []model.ServiceInfo
+	out := map[string]model.ServiceInfo{}
 	for _, host := range key.Hostnames {
-		out = append(out, a.services.ByOwningWaypoint.Lookup(NamespaceHostname{
+		for _, res := range a.services.ByOwningWaypointHostname.Lookup(NamespaceHostname{
 			Namespace: key.Namespace,
 			Hostname:  host,
-		})...)
+		}) {
+			name := res.ResourceName()
+			if _, f := out[name]; !f {
+				out[name] = res
+			}
+		}
 	}
-	return out
+
+	for _, addr := range key.Addresses {
```

Comment:
Will a waypoint ever have both a hostname and an IP? Can we only loop through addresses if `len(out) == 0`?

Reply:
Yes, the default in 1.24 is to write both to the Gateway status (and the logic for the key unconditionally includes both).

Imagine you have a single waypoint, with 2 services. In theory, I can have 2 gateways -- one with hostname, one with IP.

This is all super crazy weird setup, so not sure there is any practical way this can happen